File: calc/src/main/java/console/chapter_04/Calculator.java
package console.chapter_04;

import console.chapter_04.token.Token;
import console.chapter_04.token.Tokenizer;
import console.chapter_04.type.Operand;
import console.chapter_04.type.Operation;
import console.chapter_04.type.Operator;

import java.util.List;
import java.util.Stack;

/**
 * Chapter 04. 괄호 추가하기
 * <p>
 * condition:
 * 소괄호만
 */
public class Calculator {
    public static void main(String[] args) {
        final String input = "1 + ( 1 - 5 * 1 ) + ( ( 2 - 3 ) - 3 * 4 ) + 2 * 4";

        System.out.println(calculate(input));
    }

    private static Integer calculate(final String input) {
        final List<Token> infixTokenList = Tokenizer.tokenize(input);
        final List<Token> postfixTokenList = Postfix.changeFromInfix(infixTokenList);
        final Stack<Integer> operandStack = new Stack<>();

        System.out.println(postfixTokenList);

        for (Token token : postfixTokenList) {
            if (token instanceof Operator) {
                Operator operator = (Operator) token;

                Integer right = operandStack.pop();
                Integer left = operandStack.pop();

                operandStack.push(Operation.getExpression(operator).apply(left, right));
            } else if (token instanceof Operand) {
                operandStack.push(((Operand) token).getValue());
            } else {
                throw new RuntimeException("Invalid token");
            }
        }

        return operandStack.pop();
    }
}

File: calc/src/main/delphi/Projects/Calculator.pas
unit Calculator;

interface

function Calculate(const Input: string): Integer;

implementation

uses
  System.SysUtils, System.Generics.Collections, DebugPrinter,
  Token, Tokenizer, Stack, Calculator.Types, Postfixer;

function CalculatePostfix(const PostfixTokenList: TList<TToken>): Integer;
var
  I, R, L: Integer;
  Token: TToken;
  OperatorToken: TOperatorToken;
  OperandToken: TOperandToken;
  OperandStack: System.Generics.Collections.TStack<Integer>;
begin
  OperandStack := System.Generics.Collections.TStack<Integer>.Create;

  try
    for I := 0 to PostfixTokenList.Count - 1 do
    begin
      Token := PostfixTokenList[I];
      if Token is TOperatorToken then
      begin
        OperatorToken := TOperatorToken(Token);

        R := OperandStack.Pop;
        L := OperandStack.Pop;

        OperandStack.Push(TBinaryOperation.Operation(OperatorToken.Value, L, R));
      end
      else if Token is TOperandToken then
      begin
        OperandToken := TOperandToken(Token);
        OperandStack.Push(OperandToken.Value);
      end
      else
      begin
        raise Exception.Create('Invalid token.');
      end;
    end;

    Result := OperandStack.Pop;
  finally
    OperandStack.Free;
  end;
end;

function Calculate(const Input: string): Integer;
var
  InfixTokenList: TList<TToken>;
  PostfixTokenList: TList<TToken>;
  Tokenizer: TTokenizer;
  i: Integer;
begin
  Tokenizer := TTokenizer.Create;
  InfixTokenList := Tokenizer.Tokenize(INPUT.Replace(' ', ''));
  PostfixTokenList := Postfix(InfixTokenList);

  for i := 0 to PostfixTokenList.Count -1 do
  begin
    Print(PostfixTokenList[i].ToString);
  end;

  try
    Result := CalculatePostfix(PostfixTokenList);
  finally
    PostfixTokenList.Free;
    InfixTokenList.Free;
    Tokenizer.Free;
  end;
end;

end.
